src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/X86/AddCarry.PlatformNotSupported.cs

```diff
@@ -0,0 +1,41 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.Intrinsics.X86
+{
+    /// <summary>
+    /// This class provides access to Intel ADDCARRY hardware instructions via intrinsics
+    /// </summary>
+    [CLSCompliant(false)]
+    public abstract class AddCarry : Sse42
+    {
+        internal AddCarry() { }
+
+        public new static bool IsSupported { [Intrinsic] get { return false; } }
+
+        public new abstract class X64 : Sse41.X64
+        {
+            internal X64() { }
+
+            public new static bool IsSupported { [Intrinsic] get { return false; } }
+
+            /// <summary>
+            /// unsigned char _addcarry_u64 (unsigned char c_in, unsigned __int64 a, unsigned __int64 b, unsigned __int64 * out)
+            ///   ADC reg64, reg64
+            /// This intrinisc is only available on 64-bit processes
+            /// </summary>
+            public static unsafe byte AddWithCarry(byte carryIn, ulong left, ulong right, ulong* result) { throw new PlatformNotSupportedException(); }
```

Comment:
`byte carryIn` should be `bool carry`. It can only ever be `0` or `1` so the additional values don't matter. Likewise, it should return `bool`.

Reply:
Actually it should probably be `int`/`uint`. Dragging small int types into this will likely just make things worse for the JIT. But anyway, it's pointless to discuss this until the actual JIT implementation is designed.

Reply:
Wont `int`/`uint` make it harder to optimize against the various `cc` instructions (`setcc`, `movcc`, `jcc`, etc)?

There is also the API side of things, where we have exposed other intrinsics that "return a flag" as returning a `bool`, so the API review will likely push towards consistency.

Reply:
> Wont int/uint make it harder to optimize against the various cc instructions (setcc, movcc, jcc, etc)?

I'm not sure what `adc` has to do with `setcc`.

> There is also the API side of things, where we have exposed other intrinsics that "return a flag" as returning a bool, so the API review will likely push towards consistency.

I'd rather not comment about API review, consistency, common sense, technical realities...

Reply:
> I'm not sure what adc has to do with setcc

It is an instruction that operates on and returns the `carry` flag; so it will likely also need to plug into the instructions (like `jcc`, `cmovcc`, and `setcc`) which have specialized encodings based on the status of a given flag.

That is, if a user decided to do `if (AddWithCarry(isCarry, left, right, pResult) { /* code */ }` we should still recognize that we can just do:
```asm
add isCarry, -1
mov temp, left
adc temp, right
mov [pResult], temp
jc /* code */
```

Likewise, the simple codegen for something like `isCarry = AddWithCarry(isCarry, left, right, pResult)` (which will appear in a loop that only contains a single `AddWithCarry`) would end up being:
```asm
add isCarry, -1
mov temp, left
adc temp, right
mov [pResult], temp
setc isCarry
```

The above is all "trivially" supported in the JIT today, without many changes as it just follows similar patterns we already have.

The more complicated part ends up being producing good codegen for something like the below (which would appear, for example, in a partially unrolled loop):
```csharp
isCarry = AddWithCarry(isCarry, left[i], right[i], pResult + i);
isCarry = AddWithCarry(isCarry, left[i + 1], right[i + 1], pResult + i + 1);
```

That requires the JIT to have better understanding of what instructions modify the various flags so it knows when it can elide or delay saving the flag state until later. This is the bit that will be really difficult to handle and require more cross cutting changes to the JIT.

Reply:
> That is, if a user decided to do if (AddWithCarry(isCarry, left, right, pResult) { /* code */ } we should still recognize that we can just do:

So what you are actually trying to say that it would be useful to the return type to be `bool` so it can be easily used in an `if` statement condition. That might be the case but at the same time `bool` might cause issues when multiple additions are chained together and that would probably the most common/useful scenario. Besides, you can always use it with `if` at the "cost" of an extra ` !=0 ` that the JIT should not have any problem eliminating.